pkg/pool-stable-phantom/contracts/StablePhantomPool.sol

```diff
@@ -442,151 +442,134 @@ contract StablePhantomPool is IRateProvider, BaseGeneralPool, ProtocolFeeCache {
         uint256 indexIn,
         uint256 indexOut,
         uint256[] memory scalingFactors
-    ) private returns (uint256) {
+    ) private returns (uint256 result) {
         _upscaleArray(balances, scalingFactors);
 
         (uint256 virtualSupply, uint256[] memory balancesWithoutBpt) = _payProtocolFeesBeforeJoinExit(balances);
 
-        bool bptIsTokenIn = swapRequest.tokenIn == IERC20(this);
+        bool isJoinSwap = swapRequest.tokenOut == IERC20(this);
+        bool isGivenIn = swapRequest.kind == IVault.SwapKind.GIVEN_IN;
+
+        uint256 upscaledAmount = _upscale(swapRequest.amount, scalingFactors[isGivenIn ? indexIn : indexOut]);
         (uint256 amp, ) = _getAmplificationParameter();
 
         // The lower level function return values are still upscaled, so we need to downscale the final return value
-        if (swapRequest.kind == IVault.SwapKind.GIVEN_IN) {
-            // Returning amountOut; tokens are leaving the Vault, so we round down
-            return
-                _downscaleDown(
-                    _onSwapBptGivenIn(
-                        _upscale(swapRequest.amount, scalingFactors[indexIn]),
-                        indexIn,
-                        indexOut,
-                        bptIsTokenIn,
-                        amp,
-                        virtualSupply,
-                        balancesWithoutBpt
-                    ),
-                    scalingFactors[indexOut]
-                );
+        if (isJoinSwap) {
+            uint256 indexInNoBpt = _skipBptIndex(indexIn);
+            uint256 unknownAmount = _onSwapBptJoin(
+                upscaledAmount,
+                indexInNoBpt,
+                isGivenIn,
+                amp,
+                virtualSupply,
+                balancesWithoutBpt
+            );
+
+            if (isGivenIn) {
+                balancesWithoutBpt[indexInNoBpt] += upscaledAmount;
+                // If join is "given in" then `unknownAmount` is the amount of BPT received, so we round down.
+                result = _downscaleDown(unknownAmount, scalingFactors[indexOut]);
```

Comment:
The comment is incorrect as we send BPT in a join, not receive. 

Regardless, the logic is sound: in a 'given in' join, the amount given (here called `upscaledAmount`) is an amount in (and hence added to the balances array), and the amount calculated (here called `unknown`) is an amount out (and hence we downscale down).